pkg/metrics/metrics.go

```diff
@@ -0,0 +1,190 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"k8s.io/apimachinery/pkg/types"
+	k8smetrics "k8s.io/component-base/metrics"
+	"k8s.io/klog"
+)
+
+const (
+	labelDriverName               = "driver_name"
+	labelOperationName            = "operation_name"
+	labelOperationStatus          = "operation_status"
+	snapshotController            = "snapshot_controller"
+	operationLatencyMetricName    = "operation_total_seconds"
+	operationLatencyMetricHelpMsg = "Total number of seconds spent by the snapshot controller for an operation to complete end to end"
+)
+
+type OperationState string
+
+const (
+	// InterimFailure is an operation state which means the controller will keep
+	// on retrying the operation upon encountering this type of error.
+	InterimFailure OperationState = "InterimFailure"
+
+	// TerminatingFailure is an operation state which means the controller has encountered
+	// an error which it's not able to resume the operation and has marked the
+	// permanently failure of the operation.
+	TerminatingFailure OperationState = "TerminatingFailure"
+
+	// Success states that the controller has successfully executed the operation.
```

Comment:
Is this time to cut, or time to upload?

Reply:
it's up to the caller to decide. For volume snapshot, it should be when a content has been marked by ready to use = uploading finished.

Reply:
Do you think we should have 2 different metrics? One for time to cut and one for time to upload?

Reply:
I am not sure about that. time to upload applies to only cloud providers. For those storage systems doing only inplace snapshots, it really does not apply. WDYT?

Reply:
We specifically call out creation vs readyToUse in the API. It would be good to be explicit what this metric is measuring.

Reply:
Seems good to show both.  We could decide if the snapshot is cut by checking if the creation timestamp is set, and use readyToUse to decide if it is ready to use.  I'd avoid using "upload" in the general metrics term as not every driver supports upload so it could cause confusion.

Reply:
I buy the arguments of separating cut and uploading latency, my concern is uploading is specific to limited providers and does not really make much sense to report for the rest(a snapshot will be marked as read to use right after a success create snapshot call has been made). Thinking about how snapshot controller would try to report metrics, it reacts to update of a VolumeSnapshotContent status change, and update its corresponding VolumeSnapshot's status, and then reports a latency metric. For those providers who do not need an upload phase, the ReadyToUse flag will be true, and a metric should be reported but which one? do we report both cut/upload?

Reply:
or maybe we say we only record ready latency metrics for those providers which do not have an upload phase.

Reply:
How do you determine this?  By checking if both creation timestamp and readyToUse are returned at the same time?

Reply:
could be something as following:
if timestamp != nil {
   if readytouse {
       report readytouse metric
       return
   }
   report cut metric
}

Reply:
My suggestion is to have 2 operations, one for cut and one for upload. For the providers that don't upload, then their upload latency will be 0 and always successful. That seems fine to me.

Reply:
yeah, that's the approach I am looking to implement in controller

Reply:
Sounds good to me.